core/src/main/java/org/mskcc/oncotree/utils/CacheUtil.java

```diff
@@ -1,78 +1,114 @@
+/*
+ * Copyright (c) 2016-2018 Memorial Sloan-Kettering Cancer Center.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY, WITHOUT EVEN THE IMPLIED WARRANTY OF
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.  The software and
+ * documentation provided hereunder is on an "as is" basis, and
+ * Memorial Sloan-Kettering Cancer Center
+ * has no obligations to provide maintenance, support,
+ * updates, enhancements or modifications.  In no event shall
+ * Memorial Sloan-Kettering Cancer Center
+ * be liable to any party for direct, indirect, special,
+ * incidental or consequential damages, including lost profits, arising
+ * out of the use of this software and its documentation, even if
+ * Memorial Sloan-Kettering Cancer Center
+ * has been advised of the possibility of such damage.
+*/
 package org.mskcc.oncotree.utils;
 
-import org.mskcc.oncotree.error.InvalidOncoTreeDataException;
-import org.mskcc.oncotree.model.MainType;
+import javax.annotation.PostConstruct;
+
+import org.mskcc.oncotree.error.InvalidVersionException;
 import org.mskcc.oncotree.model.TumorType;
 import org.mskcc.oncotree.model.Version;
+import org.mskcc.oncotree.topbraid.TopBraidException;
+import org.mskcc.oncotree.utils.VersionUtil;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
+import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+
+import java.time.ZonedDateTime;
 import java.util.*;
 
 /**
  * Created by Hongxin on 2/25/16.
  */
+@Component
+@EnableScheduling
 public class CacheUtil {
-    public static Map<Version, List<MainType>> mainTypes = new HashMap<>();
-    public static Map<Version, Map<String, TumorType>> tumorTypes = new HashMap<>();
 
-    public static void resetTumorTypesByVersion(Version version) {
-        if (tumorTypes.containsKey(version)) {
-            tumorTypes.get(version).clear();
-        } else {
-            tumorTypes.put(version, new HashMap<String, TumorType>());
+    private static final Logger logger = LoggerFactory.getLogger(CacheUtil.class);
+
+    private static Map<Version, Map<String, TumorType>> tumorTypes = null;
+    private static Date lastSuccessfullRefresh = null;
+
+    @PostConstruct // call when constructed
+    @Scheduled(cron="0 0 11 * * *") // reset at 11am every day
+    private void scheduleResetCache() {
+        // TODO make sure we don't have two scheduled calls run simultaneously
+        try {
+            resetCache();
```

Comment:
Here we try to reset the cache once every day at 11. This gives only 3 chances to get a successful cache reset. Alternatively we could check more frequently but only attempt a reset if the age of the current cache is over a certain age. This would perform more retries if a reset failed.
Also, once the cache has fully expired, it would be good to have another scheduled attempt to reset the cache. Maybe a function which runs every minute (or every few minutes) and only tries to reset the cache if it is currently expired.